### ISSUE TITLE
Cache look-up for functions with optional arguments fails ...

### ISSUE THREAD

**Issue by asodeur** [numba/numba#4959]
... for non-singleton default arguments (ie booleans and small ints seem ok, floats fail).

```python
from numba import jit

DEFAULT = 123456789

@jit(nopython=True, cache=True)
def x_or_default(x=DEFAULT):
    return x


x_or_default()

# looks like the compile result gets written into a .nbc file
overloads = x_or_default._cache._cache_file._load_index()
print('Cached overloads: ', overloads)

# but you cannot load the compile result from cache
assert x_or_default._cache.load_overload(
        (types.Omitted(DEFAULT),), x_or_default.targetctx
    ) is not None

# the cause seems to be that the unpickled Omitted(DEFAULT) does not equal
# the newly created one, ie he following fails
assert list(overloads)[0][0][0] == types.Omitted(DEFAULT)
assert list(overloads)[0][0][0] == list(
            x_or_default._cache._cache_file._load_index())[0][0][0]
```
The code above will work for `DEFAULT=123` (likely due to Python's  small int caching)

**Comment by asodeur** [comment]
The issue seems to be with the definition of `Omitted.key`:
````python
@property
def key(self):
   return type(self.value), id(self.value)
```
To get the above working you had to return `type(self.value), self.value`. Hopefully, that does not break things elsewhere.

**Comment by stuartarchibald** [comment]
Thanks for the report. I can reproduce, even the round trip fails:
```python
from numba import jit
from numba import types

DEFAULT = 123456789

@jit(nopython=True, cache=True)
def x_or_default(x=DEFAULT):
    return x


x_or_default()

# looks like the compile result gets written into a .nbc file
overloads = x_or_default._cache._cache_file._load_index()
print('Cached overloads: ', overloads)

# Round trip fails
key = [_ for _ in overloads][0]
assert x_or_default._cache.load_overload(key, x_or_default.targetctx)
```

**Comment by njriasan** [comment]
@stuartarchibald is this issue resolved by #7422? Can this issue be closed?

**Comment by stuartarchibald** [comment]
> @stuartarchibald is this issue resolved by #7422? Can this issue be closed?

@njriasan I think so, OP now works as expected. Thanks
